### DISCUSSION TITLE
spring security

### DISCUSSION THREAD

**Discussion by HelloWwtABC** [community/community#50071] [Other Feature Feedback, Questions, & Ideas]
I'm a newbie and recently learned Spring Security, and I've had a problem with database-based authentication that I haven't been able to solve

1. first controller

`@RestController
@RequestMapping("/admin")
public class AdminController {
    @RequestMapping("/1")
    public String test1(){
        return "admin page info!!!";
    }
}
`

2. second controller
`@RestController
@RequestMapping("/index")
public class IndexController {
    @RequestMapping("/1")
    public String test1(){
        return "index page info!!!";
    }
}
`
3. SecurityConfiguration.java
`@Configuration
@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .antMatchers("/index/**").hasRole("USER")
                .antMatchers("/admin/**").hasRole("ADMIN")
                .anyRequest().authenticated()
                .and()
                .formLogin();
    }

    @Autowired
    private UserDetailsService userDetailsService;

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        auth.userDetailsService(userDetailsService)// 设置自定义的userDetailsService
                .passwordEncoder(passwordEncoder());

    }

    @Bean
    public PasswordEncoder passwordEncoder() {
      
        return NoOpPasswordEncoder.getInstance();
    }

    @Override
    public void configure(WebSecurity web) throws Exception {
        web.ignoring().antMatchers(HttpMethod.OPTIONS, "/**");
    }
}`
4. MyUserDetailsService.java
`@Component("userDetailsService")
public class MyUserDetailsService implements UserDetailsService {
    @Autowired
    private UserRepository userRepository;

    private GrantedAuthority DEFAULT_ROLE = new SimpleGrantedAuthority("USER");

    @Override
    public UserDetails loadUserByUsername(String loginUsername) throws UsernameNotFoundException {
        //1.根据用户名从数据库查询用户信息
        User loginUser = userRepository.findOneByLogin(loginUsername);
        //2.判断用户是否存在
        if(loginUser == null){
            System.out.println("============User" + loginUsername + "was not found in database=========");
            throw new UsernameNotFoundException("User" + loginUsername + "was not found in database");
        }
        //3.赋予权限
        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
        String dbRole = loginUser.getRole();
        //4.判断该用户是否有权限。默认为”USER“
        if(StringUtils.isNullOrEmpty(dbRole)){
            grantedAuthorities.add(DEFAULT_ROLE);
        }else {
            String[] roles = dbRole.split(",");
            for (String role : roles) {
                GrantedAuthority grantedAuthority = new SimpleGrantedAuthority(role);
                grantedAuthorities.add(grantedAuthority);
            }
        }
        return new org.springframework.security.core.userdetails.User(loginUsername, loginUser.getPassword(),grantedAuthorities);
    }
}
`

5.my question
When I access either URL after a successful login, fobidden.403 appears, and to my distress, I can't find what is wrong, can someone help me solve it.
![image](https://user-images.githubusercontent.com/127606840/225170376-f726e26e-9eb2-44c3-8a02-18bbbe2516ee.png)

**Comment by chudnovskyi** [comment]
These antMatchers are looking for URLs that end with **`/`**, not for **`/1`**. 
When you try to access **`/index/1`** or **`/admin/1`**, the antMatchers are not matching and you are getting a 403 Forbidden error.

Try this:
`.antMatchers("/index/**").hasRole("USER")`
`.antMatchers("/admin/**").hasRole("ADMIN")`
